planetiler-core/src/main/java/com/onthegomap/planetiler/util/LoopLineMerger.java

```diff
@@ -0,0 +1,397 @@
+  private GeometryFactory factory = new GeometryFactory(precisionModel);
+  private double minLength = 0.0;
+  private double loopMinLength = 0.0;
+
+  private static double getLength(List<Edge> lines) {
+    double result = 0.0;
+    for (var line : lines) {
+      result += line.length;
+    }
+    return result;
+  }
+
+  static boolean hasPointAppearingMoreThanTwice(List<Edge> edges) {
+    HashMap<Node, Integer> pointCountMap = new HashMap<>();
+    for (Edge edge : edges) {
+      if (pointCountMap.merge(edge.from, 1, Integer::sum) > 2) {
+        return true;
+      }
+      if (pointCountMap.merge(edge.to, 1, Integer::sum) > 2) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  public LoopLineMerger setPrecisionModel(PrecisionModel precisionModel) {
+    this.precisionModel = precisionModel;
+    factory = new GeometryFactory(precisionModel);
+    return this;
+  }
+
+  public LoopLineMerger setMinLength(double minLength) {
+    this.minLength = minLength;
+    return this;
+  }
+
+  public LoopLineMerger setLoopMinLength(double loopMinLength) {
+    this.loopMinLength = loopMinLength;
+    return this;
+  }
+
+  public void add(Geometry geometry) {
+    geometry.apply((GeometryComponentFilter) component -> {
+      if (component instanceof LineString lineString) {
+        input.add(lineString);
+      }
+    });
+  }
+
+  private void merge() {
+    for (var node : output) {
+      if (node.getEdges().size() == 2) {
+        Edge a = node.getEdges().getFirst();
+        Edge b = node.getEdges().get(1);
+        node.getEdges().clear();
+        List<Coordinate> coordinates = new ArrayList<>();
+        coordinates.addAll(a.coordinates.reversed());
+        coordinates.addAll(b.coordinates.subList(1, b.coordinates.size()));
+        Edge c = new Edge(a.to, b.to, coordinates, a.length + b.length);
+        a.to.removeEdge(a.reversed);
+        b.to.removeEdge(b.reversed);
+        a.to.addEdge(c);
+        b.to.addEdge(c.reversed);
+      }
+    }
+  }
+
+  private void removeLoops() {
+    for (var node : output) {
+      for (var edge : List.copyOf(node.getEdges())) {
+        var allPaths = findAllPaths(edge.from, edge.to, loopMinLength);
+        if (allPaths.size() > 1) {
+          for (var path : allPaths.subList(1, allPaths.size())) {
+            for (var toRemove : path) {
+              toRemove.remove();
+            }
+          }
+        }
+      }
+    }
+  }
+
+  List<List<Edge>> findAllPaths(Node start, Node end, double maxLength) {
+    List<List<Edge>> allPaths = new ArrayList<>();
+    Queue<List<Edge>> queue = new LinkedList<>();
+
+    for (var edge : start.getEdges()) {
+      List<Edge> path = new ArrayList<>();
+      path.add(edge);
+      if (maxLength > 0.0) {
+        if (edge.length <= maxLength) {
+          queue.add(path);
+        }
+      } else {
+        queue.add(path);
+      }
+    }
+
+    while (!queue.isEmpty()) {
```

Comment:
There are a few more potential optimizations to findAllPaths for the case where maxLength is long...

1. only compute length of currentPath once in , then just add edge.length to it
2. check if currentPathLength + edge.length + edge.to.distance(end) > max length to avoid even considering paths that can't possibly be less than max length
3. use a set of visited edges to avoid re-processing the same edge - I'm not sure if this still preserves the original intent of the function?

But I'm also not sure if it makes sense to be removing loops with a very long min length?

Reply:
Very long loops like 20 pixels long will, if the machine actually can go into all the branching and find all paths, lead to results that maybe you don't expect from the algorithm. When drawing lines in MapLibre we often use linewidths of a few pixels and as long as loops  shorter than the MapLibre `line-width` are remove, you should not really notice it. This is the goal here basically.
I think it would be best to expose a parameter in mergeLineStrings for the `loopMinLength`.